### ISSUE TITLE
runtime: plan9/amd64 builder hangs

### ISSUE THREAD

**Issue by psilva261** [golang/go#61575]
https://build.golang.org/log/ed9b4a8cec1f938b98f5cc517dcd8ca603486c91

```
ok  	regexp/syntax	0.552
Test {"go_test:runtime" "runtime"} ran over 20m0s limit (20m0.00085011s); saw output:

XXXBANNERXXX:Test execution environment.
# GOARCH: amd64
# CPU: Intel Core Processor (Broadwell)
# GOOS: plan9
# OS Version: 2000

XXXBANNERXXX:Testing packages.
```

I've seen similar on 9front/amd64 with less than 16 GB of RAM on recent versions when running `all.rc`.

**Comment by bcmills** [comment]
(CC @golang/plan9)

**Comment by psilva261** [comment]
I think it's a deadlock because usually the runtime package can be tested in 1-2 minutes and with `all.rc` no timeout seems to be ever happen

**Comment by mauri870** [comment]
The way it is timing out without any errors or messages being printed also leads me to believe it is a deadlock. I'll try to reproduce this (have to setup plan9 agian) and try to work on a fix.

**Comment by mauri870** [comment]
Unfortunately I didn't had enough disk space to compile and run the tests to investigate this issue, build stopped half way through.  If someone else is available and willing to give it a try, please feel free to jump in and assist.

**Comment by psilva261** [comment]
Actually since https://github.com/golang/go/commit/8fa9e3beee8b0e6baa7333740996181268b60a3a 16GB RAM is needed instead of 8GB. On 9front with `all.rc` (or `run.rc`) tests cannot complete `runtime` and produce no logs. (No CPU load) But running tests directly shows out-of-memory messages:

```
=== RUN   TestGcPacer/ExceedMemoryLimit
=== PAUSE TestGcPacer/ExceedMemoryLimit
=== RUN   TestGcPacer/ExceedMemoryLimitNoGCPercent
=== PAUSE TestGcPacer/ExceedMemoryLimitNoGCPercent
=== RUN   TestGcPacer/MaintainMemoryLimit
=== PAUSE TestGcPacer/MaintainMemoryLimit
=== RUN   TestGcPacer/MaintainMemoryLimitNoGCPercent
=== PAUSE TestGcPacer/MaintainMemoryLimitNoGCPercent
=== CONT  TestMemmoveAlias
runtime.test 2413: Killed: Insufficient physical memory
runtime.test 2415: Killed: Insufficient physical memory
runtime.test 2412: Killed: Insufficient physical memory
runtime.test 2423: Killed: Insufficient physical memory
exit status: 'runtime.test 2412: Killed: Insufficient physical memory'
FAIL    runtime 50.313s
runtime.test 2609: Killed: Insufficient physical memory
runtime.test 2607: Killed: Insufficient physical memory
runtime.test 2608: Killed: Insufficient physical memory
runtime.test 2414: Killed: Insufficient physical memory
runtime.test 2417: Killed: Insufficient physical memory
runtime.test 2605: Killed: Insufficient physical memory
runtime.test 2606: Killed: Insufficient physical memory
```

(`runtime` tests with `go test -v -short` and recent patches applied)

[run.rc-8GB-RAM.log](https://github.com/golang/go/files/12209309/run.rc-8GB-RAM.log)
[run.rc-16GB-RAM.log](https://github.com/golang/go/files/12209310/run.rc-16GB-RAM.log)

**Comment by denzuko** [comment]
@mauri870  I did speak with Ori and the guys on 9gridchan about this today. Since I've already been building out a builder cluster my own use and for some business clients, I'll start tackling some of the plan9/9front arm64 issues doing a series of hackathon streams starting this weekend.

@psilva261 sounds like all.rc needs some optimizations for a start right?

**Comment by psilva261** [comment]
@denzuko Yeah probably. At the moment I can reproduce it only with 4 GB RAM without having all build files in /tmp though. (Maybe there were already other changes that make the problem less drastic) But still then the tests would hang.

I think it's multiple things at once:

* out of memory
* parent process `runtime.test ...` keeps running despite a child process being killed because of out of memory (this might be a problem with `os/exec` on plan9. That reminds me of https://github.com/golang/go/issues/60793)
* out of memory error not visible in the builder because `go test` waits for test to finish/exit to print output (eventually at least the timeout from the `go test ... archive/tar archive/zip ... runtime ...` invocation for all packages should make run.rc exit and print the error)

Here's also a screenshot of `pstree` with a `-v` included in https://github.com/golang/go/blob/e9b3ff15f40d6b258217b3467c662f816b078477/src/cmd/dist/test.go#L418

<img width="1020" alt="run rc" src="https://github.com/golang/go/assets/159059/83dcc389-4e3c-48ef-8ff8-e22f19b036eb">

**Comment by denzuko** [comment]
@psilva261 Thanks for the screenshot and details. That helps tremendously. Does sounds like a lot of this is coming down to memory constraints. I'll have to dig around but will help out as I can.
